commons/src/main/java/org/sonarsource/analyzer/commons/PatternDirectoryScanner.java

```diff
@@ -0,0 +1,60 @@
+/*
+ * SonarSource Analyzers Commons
+ * Copyright (C) 2009-2021 SonarSource SA
+ * mailto:info AT sonarsource DOT com
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+package org.sonarsource.analyzer.commons;
+
+import java.io.File;
+import java.util.ArrayList;
+import java.util.List;
+import org.apache.commons.io.FileUtils;
+import org.apache.commons.io.FilenameUtils;
+import org.apache.commons.io.filefilter.IOFileFilter;
+import org.apache.commons.io.filefilter.TrueFileFilter;
+import org.sonar.api.utils.WildcardPattern;
+
+public class PatternDirectoryScanner {
+
+  private final File baseDir;
+  private final WildcardPattern pattern;
+
+  public PatternDirectoryScanner(File baseDir, String pattern) {
+    this.baseDir = baseDir;
+    this.pattern = WildcardPattern.create(pattern);
+  }
+
+  public List<File> getMatchingFiles() {
+    final String baseDirAbsolutePath = baseDir.getAbsolutePath();
+    IOFileFilter fileFilter = new IOFileFilter() {
+
+      @Override
+      public boolean accept(File dir, String name) {
+        return accept(new File(dir, name));
+      }
+
+      @Override
+      public boolean accept(File file) {
+        String path = file.getAbsolutePath();
+        path = path.substring(Math.min(baseDirAbsolutePath.length(), path.length()));
```

Comment:
do you know why we need this check? is it just defensive programming? I don't see when it can happen

Reply:
The original [commit](https://github.com/SonarSource/sonar-python/commit/d650edab79cad414514b405a88f7f356d8fb185f#diff-0ad39d8ead8e6fb85c8667610e59d773717e8008ab608d970d11ae5dda8dfdc7) doesn't really help to understand why, but judging by the following test case, I have a feeling it's either due to unexpected behaviours on some operating systems or to handle symbolic links :

https://github.com/SonarSource/sonar-analyzer-commons/blob/dda3d0a7c077cd0cad0b2d96ac799c9acbfc9dde/commons/src/test/java/org/sonarsource/analyzer/commons/PatternDirectoryScannerTest.java#L56-L71